Extract stream_subscription.py.

showell opened the pull request:
This module removes some code duplication and optimizes the stream deactivation path a bit.

Long term I'm hoping we can have something like a StreamSubscription model that uses stream_id, not recipient_id, as a foreign key.  The main goal is to remove some cognitive load for the programmer, and we'd also probably save some DB/cache hops or table joins in some places of the code (especially if we can also denormalize Message to have a stream_id).

timabbott commented:
I merged all but the last commit.  I'm a little uncomfortable with that one, mostly because we've been pretty happy with that old code for `bulk_remove_subscriptions`, but I don't actually see any sense in which the new code will be worse.  

Did you look at how carefully our existing tests verify the stream removal behavior?  

For the last commit, it seems like `get_bulk_stream_subscriber_info` would be happier being passed in the `stream_dict` structure (and then computing `stream_ids` from that), just for a bit more simplicity.

(BTW, all the new functions that return a `QuerySet` in the mypy annotations: can those return a `QuerySet[UserProfile]` (e.g.) instead?  That would avoid regressing the typing situation in the future... right now QuerySet is an Any anyway, but if we add Django stubs, QuerySet[UserProfile] would mean more than QuerySet in the same way that List[UserProfile] means more than `List`).  Definitely not worth adjusting this in the original; we can just add a change on top.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2273,27 +2274,33 @@ def notify_subscriptions_removed(user_profile, streams, no_log=False):
+    stream_ids = {stream.id for stream in streams}
+    stream_dict = {stream.id: stream for stream in streams}
+
+    info = get_bulk_stream_subscriber_info(users, streams)
```
I kinda want a more descriptive variable name than `info` here.  Maybe `subscriber_info` is good enough?

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2273,27 +2274,33 @@ def notify_subscriptions_removed(user_profile, streams, no_log=False):
+        for item in user_info:
+            sub = item['sub']
+            stream = item['stream']
+            subs_to_deactivate.append((sub, stream))
```
I kinda want more descriptive variables for `info` and `user_info`.  If we were to move the data structure to be a `(sub, stream)` tuple, I guess we could do just do `sub_tuple` and `user_sub_tuples`, or even just `for (sub, stream) in user_sub_tuples`.  Do you think that'd be more readable?

Or if we don't like the tuple idea, `sub_dict` and `user_sub_dicts` might be reasonable.

showell commented:
So one quick response is that the bulk-remove code has pretty good test coverage (beyond just line coverage).  If you comment out lines or hard code lists to empty, you reliably get failures.

showell commented:
Ok, I re-pushed.  I think I've addressed most of your feedback.  For the main commit (the second one), I followed your suggestion of using tuples, which I think does help a bit, plus I renamed some vars and made an inner function for the never-subscribed logic.

timabbott commented:
Merged, after fixing a type in a variable name.  Thanks @showell!
